### PR TITLE
Optimize predicate with toYear converter

### PR DESCRIPTION
**Pull request by ZhiguoZh**
The date converters, such as toYear, are widely used in the where clauses of the SQL queries, however, these conversions are often expensive due to the complexity of the calendar system.

The function preimage is found an optimization for the predicates with the converters. Given a predicate, toYear(c) = y, we could convert it to its equivalent form: c >= b AND c <= e, where b is "y-01-01" and e is "y-12-31". The similar transformation applies to other comparisons (<>, <, >, <=, <=).

This commit implemented the above transformation at the AST level by adding a new pass in the TreeOptimizer and a new AST visitor for in-place replacing the predicates of toYear with the converted ones.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
As is addresed in issue #49748, the predicates with date converters, such as **toYear, toYYYYMM**, could be rewritten with the equivalent date (YYYY-MM-DD) comparisons at the AST level. And this transformation could bring performance improvement as it is free from the expensive date converter and the comparison between dates (or integers in the low level representation) is quite low-cost. The [prototype](https://github.com/ZhiguoZh/ClickHouse/commit/c7f1753f0c9363a19d95fa46f1cfed1d9f505ee0) shows that, with all identified date converters optimized, the overall QPS of the 13 queries is enhanced by **~11%** on the ICX server  (Intel Xeon Platinum 8380 CPU, 80 cores, 160 threads).

This commit optimizes the predicates with **toYear** solely while taking different comparisons (=, <>, <, <=, >, >=) and order of comparitor's arguments (for example, toYear(c)<1993 and 1993>toYear(c)) into account, as the functional test suggests. It could be observed from the following figure that, with **6 of all 13** queries impacted, this optimization could improve the **overall QPS** of SSB workloads by **8.7%** and some of them could even get enhanced by **~30%**.

image
Fig 1. The performance impact of this change on SSB

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 5e83e197cd76ead5252bb140cbe8785679fb7458 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50062/5e83e197cd76ead5252bb140cbe8785679fb7458/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/OptimizeDateFilterVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+namespace DB
+{
+
+ASTPtr generateOptimizedDateFilterAST(const String & comparator, const String & converter, const String & column, UInt64 compare_to)
+{
+    const DateLUTImpl & date_lut = DateLUT::instance();
+
+    if (converter != "toYear") return {};
+
+    UInt64 year = compare_to;
```
Comment (hanfei1991) [inline src/Interpreters/OptimizeDateFilterVisitor.cpp:19]: Seems we don't need `compare_to` in the following code. Why not name the param `year`?

File: src/Interpreters/OptimizeDateFilterVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    String from_date = date_lut.dateToString(date_lut.makeDayNum(year, 1, 1));
```
Comment (hanfei1991) [inline src/Interpreters/OptimizeDateFilterVisitor.cpp:20]: maybe `start_date` / `end_date` will be clearer

**Review by hanfei1991 — APPROVED** [review]

**Comment by hanfei1991** [comment]
pretty clean code, only some naming suggestions

File: src/Interpreters/OptimizeDateFilterVisitor.cpp (hunk 1 above)
Comment (ZhiguoZh) [reply src/Interpreters/OptimizeDateFilterVisitor.cpp]: As is shown in this [prototype](https://github.com/ZhiguoZh/ClickHouse/commit/c7f1753f0c9363a19d95fa46f1cfed1d9f505ee0), `compare_to` is declared for a later optimization of `toYYYYMM`, where `compare_to` will be a constant, yyyymm, for the comparison with `toYYYYMM`. 

But I agree with you that naming `compare_to` as `year` makes much more sense in this PR. Revised accordingly, thanks for your suggestion!

File: src/Interpreters/OptimizeDateFilterVisitor.cpp (hunk 2 above)
Comment (ZhiguoZh) [reply src/Interpreters/OptimizeDateFilterVisitor.cpp]: Exactly, revised accordingly. Thanks!

**Comment by ZhiguoZh** [comment]
Thanks @hanfei1991 for reviewing the code and accepting this PR! I've added one more commit for revising this naming issue, while keeping the previous one untouched.

**Pull request merged**